### DISCUSSION TITLE
手摸切信息

### DISCUSSION THREAD

**Discussion by qianyexi** [Equim-chan/Mortal#21] [Q&A] [answered]
当玩家摸到一张手中已有的一张牌，mortal给出的q_values没有对这两张牌加以区分，是否意味着mortal没有实现学习手摸切的信息？

**Comment by Equim-chan** [comment, marked as answer]
有具体的例子吗？

**Reply by qianyexi** [reply]
https://github.com/Equim-chan/Mortal/blob/main/log-viewer/index.example.html

line 25

![image](https://user-images.githubusercontent.com/59384113/188303927-2083ddf1-e9ff-4810-a82a-339f5a07ba1b.png)

这里对家摸了一张东，手里也有一张东，mortal给出了11个合法q_values分别对应[4m, 8m, 1p, 3p, 4s, 7s, 8s, E, S, N, C]
如果考虑手摸切，摸到的这张东和手里的这张东应该有细微差别。

**Reply by Equim-chan** [reply]
是说摸切东和手切东应该作为两个不同的动作吗？

**Reply by qianyexi** [reply]
是这个意思

**Reply by Equim-chan** [reply]
我认为没有这个必要，这只会徒增输出的维度(可能变得更难 train)。其他的 AI 也没有对这个做区分，都是默认摸切，无法摸切就手切。

**Comment by wongsingfo** [comment]
Mortal 总是优先摸切。实战数据表明，默认摸切对性能影响应当不大。我对凤桌[1]约 100K 场半庄进行了统计，其中包含了 2143169 次手摸切的判断，其中 91.34% 选择了摸切。

另外，值得一提的是，当需要鸣牌时，Mortal 的动作空间也不区分是否要使用赤牌，而是“贪心”地优先使用赤牌。比如，手上有五万和红五万，上家打出三四万，如果 Mortal 决定了要吃，那么总是使用红五万去吃。凤桌的数据统计表明，当可以同时选择使用赤牌鸣牌和不使用赤牌鸣牌时，97.78% 的决策使用了赤牌鸣牌。这表明目前没有实现区分赤牌鸣牌的必要。

[1] https://tenhou.net/ranking.html

**Reply by qianyexi** [reply]
感谢说明
